Feature/add migrations

anxolin opened the pull request:
Dima, would you help me finishing this PR?

- This is our migration scripts
- Merged with the PK support
- Merged with the first DAO deployment to Rinkeby
- (and this was was the pain) Merged with some changes we had unmerged in master

It looks good and migrations work. But it doesn't pass the TESTs.

Also, we can add here the changes so we pull the addresses of MGN, DX, WETH, GNO from the dependant projects. Maybe for GNO and WETH is not worth it and we can just edit the `networks.json` file and add the addresses for the 3 networks: `kovan`, `rinkeby`, `mainnet`

anxolin commented:
Hei @orenyodfat !

Please review this PR, it has the initial version of the deployment. 
We still have to review a bit more the config, but I think is a very good PR.

After this one, I'll make a second PR with the latest version of ARC and the price feed, already deployed and with some data.

Thanks!

orenyodfat commented on contracts/DxAvatar.sol:
```diff
@@ -0,0 +1,11 @@
+pragma solidity ^0.4.24;
```
how about 0.0.25 as in arc ?

anxolin replied:
I thought arc was `0.4.24`. Anyways, we 0.4.25 there's a problem. Some projects are not yet migrated to truffle 5, because as you know is a big change and tests and JS code must be adapted. 

So the solution to be able to have both versions of truffle working a the same time with the same contracts we need to use 0.4.24 (last version used by truffle 4)

For the ones in the DAO we could use 0.4.25, cause we use truffle 5, but I wonder if this would be a limitation while people is not fully migrated. What do you think?
Is there something important in 0.4.25 that would make it worth it?

orenyodfat replied:
arc use 0.4.25.
in order to be able to verify the bytecode (against arc) it need to be compile with the same compiler version.

anxolin replied:
Yep, I understand. But, I mean, is not wise to consider using `0.4.24` since has more compatibility.

We can adapt, it's just, we could weight the pros/cons

anxolin replied:
Actually, I think we use the 0.4.25 compiler, but the pragma 0.4.24. Is not this a fare setup?

anxolin replied:
It doesn't solve the issue, does it? 
I think https://www.npmjs.com/package/truffle/v/4.1.14 uses 0.4.24 compiler. I understand that if we try to compile a contract with pragma 0.4.25 will fail, right?

Or am I understanding it wrong?

orenyodfat replied:
use truffle https://github.com/trufflesuite/truffle/releases/tag/v5.0.0-beta.2

orenyodfat commented on contracts/DxToken.sol:
```diff
@@ -0,0 +1,11 @@
+pragma solidity ^0.4.24;
```
0.0.25 ?

anxolin replied:
Same as the other comment

anxolin commented:
@orenyodfat We'll consolidate ALL in development branch while the PR is not approved. Anyways, do you think next week you can validate it. It's important for us to merge this to master and sync here.

orenyodfat commented on .eslintrc.json:
```diff
@@ -12,6 +12,11 @@
+        "node": true
     }
-}
+}
```
please remove CR

anxolin replied:
ok

orenyodfat commented on .gitignore:
```diff
@@ -1,4 +1,5 @@
-build
\ No newline at end of file
+build
+.env
```
please remove CR

anxolin replied:
ok

orenyodfat commented on contracts/DxAvatar.sol:
```diff
@@ -0,0 +1,11 @@
+        DAOToken _nativeToken,
+        Reputation _nativeReputation
+    ) Avatar(_orgName, _nativeToken, _nativeReputation) public {}
+}
```
please remove CR

anxolin replied:
ok

orenyodfat commented on contracts/DxReputation.sol:
```diff
@@ -0,0 +1,8 @@
+
+import "@daostack/infra/contracts/Reputation.sol";
+
+// is DAOToken
```
DAOToken ?

anxolin replied:
is the reputation, corrected

orenyodfat commented on test/contracts/MgnToken.sol:
```diff
@@ -0,0 +1,177 @@
+    /// @param a First addend
+    /// @param b Second addend
+    /// @return Did no overflow occur?
+    function safeToAdd(uint a, uint b)
```
why not use zeppelin safe math ?

anxolin replied:
Cause that's a copy of the original MgnToken. We wanted to use the same logic. Just for testing.

We do plan to change to Safe Math at some point, but is not the porpoise of this project.

anxolin replied:
Anyways, I just delete the file cause we use it to do some tests. But now we use the original MGN

orenyodfat commented on test/contracts/MgnToken.sol:
```diff
@@ -0,0 +1,177 @@
+import "./TestToken.sol";
+
+// TODO: Review also if we can use the real MGN easily
+contract TokenFRT is TestToken {
```
why not test with the real MgnToken contract ?

anxolin replied:
We do, this files is not used anymore. I just deleted it.

orenyodfat commented:
Does this PR cover also the genericAction scheme functionality  ? acting as the dutchX auctioneer ?

orenyodfat commented on package.json:
```diff
@@ -1,42 +1,58 @@
-    "@daostack/arc": "^0.0.0-alpha.52",
-    "@gnosis.pm/dx-contracts": "^1.0.3",
-    "debug": "^3.1.0",
+    "@daostack/arc": "0.0.0-alpha.56",
```
use arc 0.0.0-alpha.58

anxolin replied:
We will update in another branch as a new PR as soon as you approve this one.

anxolin commented on contracts/LocalDevelopment.sol:
```diff
@@ -0,0 +1,21 @@
+pragma solidity ^0.4.24;
+
+// DutchX dependencies
```
Is just to make truffle compile some contracts that are not in the `/contracts` dir

anxolin commented:
> Does travis runs solium ? eslint ?

They both were configured in the project, but travis was not running the lint.
I added that in https://github.com/gnosis/dx-daostack/pull/33 

Also I fixed the lint errors for solium and eslint.

anxolin commented:
@orenyodfat
